### PR TITLE
feat: store '.so' file names in database instead of using marker files

### PR DESCRIPTION
**Pull request by bmoxb**
* Compiled '.so' files are still given random names but a service deployment can be mapped to a '.so' file name via a new `libs` database table.
* All '.so' files are moved to a common directory instead of being left in the project directory.

### CODE REVIEW

File: deployer/src/deployment/queue.rs [hunk 1]
```diff
@@ -117,53 +117,52 @@ fn extract_tar_gz_data(data: impl Read, dest: impl AsRef<Path>) -> Result<()> {
...
+async fn remove_old_lib(
+    persistence: &Persistence,
+    name: &str,
+    libs_path: impl AsRef<Path>,
+) -> Result<()> {
+    if let Some(lib_name) = persistence.get_last_built_lib(name).await? {
+        let lib_path = libs_path.as_ref().join(lib_name);
+
+        if lib_path.exists() {
+            fs::remove_file(lib_path).await?;
```
Comment (chesedo) [inline deployer/src/deployment/queue.rs:131]: I'm wondering, what will happen when a new deployment version fails to start during run. We will want to return to a previous instance, right?

Maybe we shouldn't delete old 'so' files?

File: deployer/src/persistence.rs [hunk 2]
```diff
@@ -45,13 +45,18 @@ impl Persistence {
                 state INTEGER          -- Enum indicating the current state of the deployment.
             );
 
+            CREATE TABLE IF NOT EXISTS libs (
+                name TEXT PRIMARY KEY, -- Name of the service.
+                lib TEXT               -- The name of linked library produced the last time the service was built.
+            );
```
Comment (chesedo) [inline deployer/src/persistence.rs:51]: Maybe we can remove this table if we give each incoming deployment an id and store the id in the deployments table? Then we can use the same id when renaming the 'so' file?

**Review by chesedo — COMMENTED** [review]
Oh nice!! This will be easier

File: deployer/src/deployment/queue.rs (hunk 1 above)
Comment (bmoxb) [reply deployer/src/deployment/queue.rs:131]: @chesedo Very point - perhaps the deletion logic should be moved into the loading/execution stage then?

File: deployer/src/persistence.rs (hunk 2 above)
Comment (bmoxb) [reply deployer/src/persistence.rs:51]: @chesedo Sorry if I'm misunderstanding, but are you saying that each deployment would then just have a single '.so' filename even after re-deployment? I don't think that would work due to our library caching/reloading issue (i.e., each new compiled '.so' file needs to have a unique name to be correctly loaded) :/

Comment (chesedo) [reply deployer/src/persistence.rs:51]: Yes, each deployment would have a single 'so' for that deployment. This of course assumes we are sticking with the terminology of a deployment being a version of a service

Comment (chesedo) [reply deployer/src/persistence.rs:51]: So technically I guess we should have 3 tables:
- Service (S) with the service name
- Deployments (D) with the deployment id. Here exists a (D) many-to-one (S) link
- Logs (L) with the logs for a deployment. Here exists a (L) many-to-one (D) link

The last two tables can always store the service name again if we want the DB view to remain easy (I guess the S table is then no longer needed really)

**Comment by chesedo** [comment]
Will be replaced by #242

**Pull request closed without merging**